signer/src/network/in_memory.rs

```diff
@@ -0,0 +1,73 @@
+//! # In-memory signer network client.
+//!
+//! The client itself is a thin wrapper over a tokio broadcast
+//! channel, with deduplication logic to prevent a single client
+//! from receiving it's own messages.
+
+use std::collections::VecDeque;
+
+use tokio::sync::broadcast;
+
+pub const BROADCAST_CHANNEL_CAPACITY: usize = 10_000;
+
+#[derive(Debug)]
+pub struct Client {
+    sender: broadcast::Sender<super::Msg>,
+    receiver: broadcast::Receiver<super::Msg>,
+    recently_sent: VecDeque<super::Msg>,
+}
+
+/// In-memory communication network
+#[derive(Debug)]
+pub struct Network {
+    sender: broadcast::Sender<super::Msg>,
+}
+
+impl Network {
+    /// Construct a new in-memory communication entwork
+    pub fn new() -> Self {
+        let (sender, _) = broadcast::channel(BROADCAST_CHANNEL_CAPACITY);
+        Self { sender }
+    }
+
+    /// Connect a new signer to this network
+    pub fn connect(&self) -> Client {
+        let sender = self.sender.clone();
+        let receiver = sender.subscribe();
+        let recently_sent = VecDeque::new();
+
+        Client {
+            sender,
+            receiver,
+            recently_sent,
+        }
+    }
+}
+
+impl super::Client for Client {
```

Comment:
I think we should call either the struct or the trait something different than `Client` (I kind of think we should rename them both). Maybe `InternalNetworkClient` for the trait and `MpmcBroadcaster` for the struct. These aren't the best suggestions, I'm just trying to help.

Reply:
Sure thing, I'm used to different conventions in different repos and I guess we'll figure out what feels best for us. 

On one hand, I think this works pretty well to reuse simple names such as `Error` or `Client` if the name space is informative. In this scenario, I think `in_memory::Client` looks pretty good in the code. I originally called it `InMemoryClient` which felt a bit repetitive with the namespacing, but works better for people who prefer to import structs by name without any namespacing information. Though I'm personally not a big fan of direct type imports, particularly because it makes it very hard to disambiguate between type imports and bringing traits into scope.

On the other hand, I also think it's ugly to have the same name for the trait and the implementation. Typically traits should be verbs (Clone, Print, Serialize, Encode etc.) - so from this convention it is the trait name that should be updated.

From this reasoning, I would propose renaming the trait to `MessageTransfer`, and keep the in memory network client type name to just `Client`.

Does that sound good to you?

Reply:
`MpmcBroadcaster` is more specific than `Client` though. I like that name, so let's use it.

TL; DR
I will rename the trait to `MessageTransfer` and the struct to `MpmcBroadcaster`.

Reply:
> On one hand, I think this works pretty well to reuse simple names such as `Error` or `Client` if the name space is informative. In this scenario, I think `in_memory::Client` looks pretty good in the code. I originally called it `InMemoryClient` which felt a bit repetitive with the namespacing, but works better for people who prefer to import structs by name without any namespacing information. Though I'm personally not a big fan of direct type imports, particularly because it makes it very hard to disambiguate between type imports and bringing traits into scope.

Ahhh, I see. Yeah I can relate, even though I'm definitely guilty of this in my code.

> On the other hand, I also think it's ugly to have the same name for the trait and the implementation. Typically traits should be verbs (Clone, Print, Serialize, Encode etc.) - so from this convention it is the trait name that should be updated.
> 
> From this reasoning, I would propose renaming the trait to `MessageTransfer`, and keep the in memory network client type name to just `Client`.
> 
> Does that sound good to you?

🙏🏽 Yeah that sounds great.

Reply:
Updated. Let me know what you think.